xwiki-rendering-api/src/main/java/org/xwiki/rendering/internal/parser/reference/GenericLinkReferenceParser.java

```diff
@@ -129,31 +129,45 @@
         return URI_PREFIXES;
     }
 
+
+    /*
+    Changes:
+    I set up a nested if statement to check if each value of resourceReference is null at each step.
+    If it is, it moves on to the next step and performs an action, then checks to see if it is still null.
+    If at the end of this all, the value of resourceReference is not null, resourceReference will be returned.
+    If it is still null, it returns a reference to a document.
+    This halves the number of return statement from 4 to 2, while only parsing when required.
+     */
     @Override
     public ResourceReference parse(String rawReference)
```

Comment:
I still find the whole thing harder to read than the previous version which had very clear and separated steps. Keep in mind that having several returns is not a Java bug, it's just a readability advice from Sonar so it should not be fixed at all cost: only if you can find a more readable version which is not the case here.